paddle/fluid/framework/variable.h

```diff
@@ -101,8 +110,48 @@ class Variable {
   };
 
   // pointers to a PlaceholderImpl object indeed.
-  std::unique_ptr<Placeholder> holder_;
+  std::shared_ptr<Placeholder> holder_;
 };
 
+inline framework::TensorInplaceVersion& Variable::InplaceVersionCounter() {
+  if (IsType<framework::Tensor>()) {
+    return GetMutable<framework::Tensor>()->InplaceVersionCounter();
+  } else if (IsType<framework::LoDTensor>()) {
+    return GetMutable<framework::LoDTensor>()->InplaceVersionCounter();
+  } else if (IsType<framework::SelectedRows>()) {
+    return GetMutable<framework::SelectedRows>()
+        ->mutable_value()
+        ->InplaceVersionCounter();
+  } else if (IsType<framework::LoDTensorArray>()) {
+    auto tensor_array = GetMutable<framework::LoDTensorArray>();
+    return tensor_array->at(0).InplaceVersionCounter();
+  } else {
+    // NOTE(liym27): Other types of data are not supported, like Scope.
+    PADDLE_THROW(platform::errors::Unimplemented(
+        "Only supports Tensor, LoDTensor, SelectedRows and "
+        "LoDTensorArray to have TensorInplaceVersion."));
+  }
+}
+inline uint32_t Variable::CurrentInplaceVersion() {
+  try {
+    return InplaceVersionCounter().CurrentVersion();
+  } catch (platform::EnforceNotMet& exception) {
```

Comment:
Why not raise the exception and stop the execution?

Reply:
Because the exception is raised in InplaceVersionCounter() when the type is not LoDTensor or SelectedRows, which is within expectations. And only return 0 as its inplace version.